### PR TITLE
Replace espree with babel.

### LINKED ISSUE #19952: docgen: Support ES2019, approved ES2020 syntax
Previously: https://github.com/WordPress/gutenberg/issues/18045#issuecomment-578161548
Related: https://github.com/WordPress/gutenberg/pull/19931#issuecomment-579847333, #19831

Docgen uses a different JavaScript parser ([Espree](https://github.com/eslint/espree)) than that which is used in the plugin build process ([Babel](https://babeljs.io/)). This hasn't been much of an issue thus far, but with especially new syntaxes, it can cause errors to be thrown when running `npm run docs:build` and with the corresponding pre-commit step.

Examples:

- [ES2019 optional catch binding](https://2ality.com/2017/08/optional-catch-binding.html)
   - Was intended to be used in #19871
- [ES2020 optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)
   - Was intended to be used in #19931

Originally, I thought it might just be a matter of keeping the `espree` dependency up to date, assuming that it's been updated for newer versions of the specification.

This is unfortunately not as simple as it might seem, since one of the breaking changes introduced in [`espree@5.0.0`](https://github.com/eslint/espree/releases/tag/v5.0.0) is to remove the `attachComment` feature, which is pretty critical to how `docgen` works (or at least is currently implemented). Thus, upgrading to the latest version would require a pretty significant refactor to the code, one where we'd likely emulate how [comment attachment](https://github.com/eslint/espree/pull/405/files#diff-474d681dc53859768e80de42f3d9a4ca) worked previously, using the [`comment` (and perhaps `tokens`) option(s)](https://github.com/eslint/espree#usage) to infer where a comment occurs in relation to an exported member. I started down this path on Friday, but it seems significant enough to warrant whether it's worth the effort, considering if the alternative to Doctrine would bring with it a different parser implementation as well (e.g. TypeScript).

One thing I noted in the process of this is that we configure the ECMAScript version here:

https://github.com/WordPress/gutenberg/blob/1fb3c386001266ad03ed6d2cd3d9ad8777090028/packages/docgen/src/engine.js#L15

This is particularly relevant for my [previous comment](https://github.com/WordPress/gutenberg/issues/18045#issuecomment-578161548) where I mention being unable to use ES2019 features. It's pretty clear by this code why this wouldn't be expected to work 😄 I haven't tested yet, but it's possible that Espree versions prior to the breaking 5.0 version might have support for newer versions of ECMAScript, so that we could still use those language features without bumping to the latest Espree version.

**Task:** Allow new syntaxes to be parsed without error by `docgen`.

**Proposed solution:**

One of:

1. Replace Doctrine and Espree with TypeScript (#18045).
2. Or: Bump Espree to the latest version, including necessary refactoring to `docgen` to support removal of `attachComment` option.
3. Or: Optionally bump Espree to the last version including `attachComment` option, and check to see whether it can support ES2019 and ES2020 as the configured `ecmaVersion` option.

**Follow-up tasks:**

Revisit the "intended to be used" examples from above, to introduce the desired syntax.

- [ ] https://github.com/WordPress/gutenberg/pull/19871/files#r372503756
- [ ] https://github.com/WordPress/gutenberg/pull/19931#issuecomment-579847333

### PR DESCRIPTION
**Pull request by sainthkh**
## Description

Replaced `espree` with `babel`. Fixes https://github.com/WordPress/gutenberg/issues/19952

## How has this been tested?

Ran `npm run docs:build` and found no change.

## Screenshots 

N/A

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] My code has proper inline documentation. 
- [N/A] I've included developer documentation if appropriate. 
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by nosolosw** [comment]
These are the testing steps I'm following;

- In `master`
  - Modify the JSDoc comment of some exported entity and use the optional chaining syntax. Suggestion: modify `getCurrentUser` in `packages/core-data/src/selectors.js`.
  - Run `npm install && npm run docs:build`.
  - The expected result is that it should fail due to the unrecognized syntax.
  - Clean up any changes.
- Switch to this branch:
  - Modify the JSDoc comment of some exported entity and use the optional chaining syntax. Suggestion: modify `getCurrentUser` in `packages/core-data/src/selectors.js`.
  - Run `npm install && npm run docs:build`
  - Expected result is that the appropiate docs will be updated. Following the suggestion above: there should be changes in `docs/designers-developers/developers/data/data-core.md` and `packages/core-data/README.md`.

As per these steps, this works as expected! Happy to see that the code changes are minimal. Going to review the code now.

File: packages/docgen/src/engine.js [hunk 1]
```diff
@@ -1,25 +1,27 @@
...
  */
-const espree = require( 'espree' );
+const babel = require( '@babel/parser' );
 const { flatten } = require( 'lodash' );
 
 /**
  * Internal dependencies.
  */
 const getIntermediateRepresentation = require( './get-intermediate-representation' );
+const babelConfig = require( '../../../babel.config' );
```
Comment (nosolosw) [inline packages/docgen/src/engine.js:11]: We shouldn't use files from outside the `packages/docgen` repository, otherwise, it'll fail when used as a standalone npm binary.

File: packages/docgen/src/engine.js [hunk 2]
```diff
@@ -1,25 +1,27 @@
...
-			jsx: true,
-		},
-		sourceType: 'module',
+const getAST = ( source ) => {
+	const { presets, plugins } = babelConfig( {
+		cache: () => {}, // To bypass api.cache is not a function error.
 	} );
 
+	return babel.parse( source || '', {
+		presets,
```
Comment (nosolosw) [inline packages/docgen/src/engine.js:19]: It doesn't look like babel-parser admits a [preset option](https://babeljs.io/docs/en/babel-parser/#options)?

File: packages/docgen/src/engine.js [hunk 3]
```diff
@@ -1,25 +1,27 @@
...
+		plugins: [ ...plugins, 'jsx' ],
+		sourceType: 'module',
+	} ).program; // unlike espree, the root of babel AST is File, not Program.
```
Comment (nosolosw) [inline packages/docgen/src/engine.js:22]: One thing that I usually do in these cases (comments that are useful during the review phase but perhaps don't make a lot of sense in the source code itself) is to add them myself as comments in the PR. Perhaps we could remove it?

**Review by oandregal — CHANGES_REQUESTED** [review]
Good progress here, this is a nice well-scoped refactor!

We'd need to figure out a way for us to:

1) use the same configuration than Gutenberg without hard-coding the path
2) use the presets already available (that include JSX parsing)

File: packages/docgen/src/engine.js (hunk 2 above)
Comment (sainthkh) [reply packages/docgen/src/engine.js:19]: Right, it should be removed.

File: packages/docgen/src/engine.js (hunk 3 above)
Comment (sainthkh) [reply packages/docgen/src/engine.js:22]: I'll do it.

File: packages/docgen/src/engine.js (hunk 2 above)
Comment (sainthkh) [reply packages/docgen/src/engine.js:19]: I remember it didn't work properly without this. I'll check it out by reading @babel/parser code.

File: packages/docgen/src/engine.js (hunk 1 above)
Comment (sainthkh) [reply packages/docgen/src/engine.js:11]: Then, we can make a command line option like --babel-config. And pass the path to the file in `docs` folder.

How about this?

**Comment by sainthkh** [comment]
After your comment about `presets` option, it made me wonder why it doesn't exist while `plugins` does. A babel `preset` is just a package of multiple `plugin`s. It's a bit weird to learn that you can set individual plugins while cannot set packages. 

After thinking about this weirdness for a while, I realized that we don't need `babel.config.js` at all.

When using babel, we expect it to transpile new syntax to old syntax. To do so, babel parses js code and runs plugins to generate new one. 

In other words, babel works in these 3 steps.

1. babel/parser parses code and generates AST.
2. plugins manipulates the AST and create new one.
3. babel generates new code from new AST.

As the jobs of babel/parser and babel plugins are different, we don't need to import `babel.config.js` at all. 

I removed other configs and removed them.

File: packages/docgen/src/engine.js [hunk 4]
```diff
@@ -1,24 +1,20 @@
...
-	espree.parse( source, {
-		attachComment: true,
-		loc: true,
-		ecmaVersion: 2018,
-		ecmaFeatures: {
-			jsx: true,
-		},
+const getAST = ( source ) => {
+	return babel.parse( source || '', {
+		plugins: [ 'jsx' ],
```
Comment (nosolosw) [inline packages/docgen/src/engine.js:14]: For testing, I'm using this code that includes a variety of new syntax (nullish coalescencing, optional chaining, optional catch binding, and numerical separator):

```js
	try {
		return state?.currentUser;
	} catch {
		return state ?? 1_000;
	}
```

try adding it to the `getCurrentUser` selector within the `packages/core-data/src/selectors.js` file and see how `npm run docs:build` fails due to unrecognizing the numerical separator syntax. The same code works fine with `npm run build`.

To be fully compatible with our current setup, we need to add the [numerical separator proposal](https://babeljs.io/docs/en/next/babel-parser#ecmascript-proposals-https-githubcom-babel-proposals) as a plugin as well:

```suggestion
		plugins: [ 'jsx', 'numericalSeparator' ],
```

The reason is that our `@wordpress/babel-preset-default` config [makes use](https://github.com/WordPress/gutenberg/blob/master/packages/babel-preset-default/index.js#L20) of the [shippedProposals](https://babeljs.io/docs/en/babel-preset-env#shippedproposals) option, which, at the time of writing, also enables the numeric separator syntax ([see](https://github.com/babel/babel/blob/v7.9.5/packages/babel-compat-data/scripts/data/plugin-features.js#L124)).

Comment (nosolosw) [reply packages/docgen/src/engine.js:14]: How about we add a comment here that says something along these lines: 

When upgrading babel, we need to make sure this list of plugins are updated to include the shippedProposals that we enable by default in `@wordpress/babel-preset-default` https://github.com/WordPress/gutenberg/blob/master/packages/babel-preset-default/index.js#L20

**Comment by nosolosw** [comment]
Good point about the transforms plugins. We do need the syntax/parser plugins for the parser to understand some syntax potentially used in Gutenberg: in our case, the `jsx` plugin as well as the `numericalSeparator` proposal (see [full discussion](https://github.com/WordPress/gutenberg/pull/21853#discussion_r415847682)).

With that change, I think we're good to go.

**Comment by nosolosw** [comment]
It does bother me a bit that we need to keep an additional babel config beyond the existing ones (wordpress/babel-preset-default + gutenberg babel.config.js). What if instead of using `babel-parser` we use `@babel/core` as we [do in Gutenberg](https://github.com/WordPress/gutenberg/blob/master/bin/packages/build-worker.js#L7). :thinking:

It looks like, that way, we could reuse `@wordpress/babel-preset-default` and won't need any extra config. If, at some point in the future, we needed also something was added to the Gutenberg's `babel.config.js` we could copy it verbatim (or even make it reusable), which sounds better than having to look into babel internals to see which proposals to enable, etc.

**Comment by sainthkh** [comment]
I tested your proposal about `@babel/core` by adding a file at `docgen/src/t.js` with the code below:

```js
const core = require( '@babel/core' );
const parser = require( '@babel/parser' );

const code = `
try {
	state?.currentUser;
} catch {
	state ?? 1_000;
}
`;

try {
	core.parse( code );
	process.stdout.write( 'core passed\n' );
} catch ( e ) {
	process.stdout.write( `core failed: ${ e }\n` );
}

try {
	parser.parse( code, {
		plugins: [ 'jsx' ],
		sourceType: 'module',
	} );
	process.stdout.write( 'parser passed\n' );
} catch ( e ) {
	process.stdout.write( `parser failed: ${ e }\n` );
}

try {
	parser.parse( code, {
		plugins: [ 'jsx', 'numericSeparator' ],
		sourceType: 'module',
	} );
	process.stdout.write( 'parser2 passed\n' );
} catch ( e ) {
	process.stdout.write( `parser2 failed: ${ e }\n` );
}
```

Confirmed that `@babel/core` works correctly without any setting. Let's continue in that way.

**Comment by nosolosw** [comment]
This is nice! We also get the Gutenberg root configs for free :clap: 

I think we're ready to merge, the only thing left would be to document how this works in the README. Perhaps we can add a section called "Babel configuration" after the "CLI Options" that says that it takes the [project-wide config](https://babeljs.io/docs/en/next/config-files#project-wide-configuration), what happens if there is none, etc. What do you think?

Also, could you test that documents are generated properly with the pre-commit hook? For some reason, the pre-commit hook is not working for me and I want to make sure the pre-commit hook doesn't mess up the current working directory babel uses to find the config.

Testing instructions would be:

- Paste the testing code within the `getCurrentUser` selector at `packages/core-data/src/selectors.js`.
- Modify some part of its JSON comment.
- `git add packages/core-data/src/selectors.js` and commit.
- The expected result is that the commit should fail because there are two new files to add (`docs/designers-developers/developers/data/data-core.md` and `packages/core-data/README.md`).

**Comment by sainthkh** [comment]
* I confirmed that pre-commit hook generates docs properly and commit fails.
* Updated documentation

**Review by oandregal — APPROVED** [review]

**Comment by nosolosw** [comment]
Good work here, @sainthkh!

**Comment by youknowriad** [comment]
Impressive work here :)

**Comment by nb** [comment]
Just wanted to say “Thank you”, @sainthkh  – I was working on some package changes including  and every commit was failing, because docgen did not supporting newer ECMASript features. Not the best experience you can imagine…

Today, I rebased with `master` and it was all fixed :)
